base/reducedim.jl

```diff
@@ -294,10 +129,216 @@ function _mapreducedim!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted)
+# We want to pick the _smallest_ split. For example, given four non-singleton dimensions, we have a choice of:
+# first: (:, :, :, 1:1)   or (:, :, 1:1,   :) or (:, 1:1, :, :)   or (1:2,   :, :, :)
+# rest:  (:, :, :, 2:end) or (:, :, 2:end, :) or (:, 2:end, :, :) or (3:end, :, :, :)
+# With dimension lengths of (a, b, c, d), that's minimizing: (a*b*c) vs (a*b*d) vs (a*c*d) vs (2*b*c*d)
+# The smallest subsection will always be the one that slices on the _longest_ dimension.
+function _split2(ci::CartesianIndices{N}) where {N}
+    inds = ci.indices
+    first_non_singleton = something(findfirst((>)(1), size(ci)))
+    # The first non-singleton dimension "costs half" because we take two elements
+    lengths = map(length, inds)
+    weights = setindex(lengths, lengths[first_non_singleton]÷2, first_non_singleton)
+    sliced_dim = argmax(weights)
+
+    nskip = Int(sliced_dim == first_non_singleton)
+    return (CartesianIndices(ntuple(d->d==sliced_dim ? inds[d][begin:begin+nskip] : inds[d][begin:end], Val(N))),
+            CartesianIndices(ntuple(d->d==sliced_dim ? inds[d][begin+nskip+1:end] : inds[d][begin:end], Val(N))))
+end
+
+_mapreduce_empty_array(_, _, init, A, axes, ::Nothing) = fill!(mapreduce_similar(A, typeof(init), axes), init)
+_mapreduce_empty_array(_, _, init, A, axes, out) = fill!(out, init)
+function _mapreduce_empty_array(f, op, ::_InitialValue, A, axes, ::Nothing)
+    init = mapreduce_empty(f, op, eltype(A))
+    return fill!(mapreduce_similar(A, typeof(init), axes), init)
+end
+function _mapreduce_empty_array(f, op, ::_InitialValue, A, axes, out)
+    init = mapreduce_empty(f, op, eltype(A))
+    return fill!(out, init)
+end
+
+_mapreduce_one_array(f, op, init, A, axes, ::Nothing) = collect_allocator(_mapreduce_similar_curried(A), (op(init, f(@inbounds(A[i]))) for i in CartesianIndices(axes)))
+_mapreduce_one_array(f, op, init, A, axes, out) = (for i in CartesianIndices(axes); out[i] = op(init, f(@inbounds(A[i]))); end; out)
+_mapreduce_one_array(f, op, ::_InitialValue, A, axes, ::Nothing) = collect_allocator(_mapreduce_similar_curried(A), (mapreduce_first(f, op, @inbounds(A[i])) for i in CartesianIndices(axes)))
+_mapreduce_one_array(f, op, ::_InitialValue, A, axes, out) = (for i in CartesianIndices(axes); out[i] = mapreduce_first(f, op, @inbounds(A[i])); end; out)
+
+# Given two initial values in a reduction, compute the beginning of a op chain
+_mapreduce_start(f, op, init, a1, a2) = op(op(init, f(a1)), f(a2))
+_mapreduce_start(f, op, ::_InitialValue, a1, a2) = op(f(a1), f(a2))
+
+"""
+    mapreduce!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted; [init])
+
+Compute `mapreduce(f, op, A; init, dims)` where `dims` are the singleton dimensions of `R`, storing the result into `R`.
+
+!!! note
+    The previous values in `R` are _not_ used as initial values; they are completely ignored
+"""
+function mapreduce!(f, op, R::AbstractArray, A::AbstractArrayOrBroadcasted; init=_InitialValue())
```

Comment:
how public is this function? I think there are some considerations similar to those touched on in https://github.com/JuliaLang/julia/pull/56673 about how much axes validation to do. consider:



```
julia> A = rand(2, 3, 4);

julia> mapreduce(identity, +, A)
14.014839387268804

julia> Base.mapreduce!(identity, +, [1.0], A)
1-element Vector{Float64}:
 14.014839387268806

julia> Base.mapreduce!(identity, +, [1.0;;], A)
1×1 Matrix{Float64}:
 14.014839387268806

julia> Base.mapreduce!(identity, +, [1.0;;;], A)
1×1×1 Array{Float64, 3}:
[:, :, 1] =
 14.014839387268806

julia> Base.mapreduce!(identity, +, [1.0;;;;], A)
ERROR: DimensionMismatch: cannot reduce (2, 3, 4) array into an array of (1, 1, 1, 1)
```